<font color='blue'>
    
# ECE449 Machine Learning - Assignment 2

## Task 3: LeNet with PyTorch


[//]: # "Notebook Created by Jinghua Wang (jinghuawang@intl.zju.edu.cn), last modified on 2020-11-04"

</font>

*LeNet* is one of the first published CNNs to capture wide attention for its performance on computer vision image classification tasks.
The model was introduced by (and named for) Yann LeCun, then a researcher at AT&T Bell Labs, for the purpose of recognizing handwritten digits in images, in the paper "Gradient-Based Learning Applied to Document Recognition" by Yann LeCun, Leon Boottou, Yoshua Bengio, and Patrick Haffner.
The paper is available online at: http://yann.lecun.com/exdb/publis/pdf/lecun-98.pdf.


## LeNet

At a high level, LeNet (LeNet-5) consists of two parts:

(i) A convolutional encoder consisting of two convolutional layers.

(ii) A dense block consisting of three fully-connected layers.

Below is the data flow in LeNet, the input is a handwritten digit, the output a probability over 10 possible outcomes:
<img src="img/lenet.png" width="1000">


We took a small liberty with the original LeNet-5 model by removing the Gaussian activation in the final layer.
Other than that, our LeNet architechture should match the original LeNet-5 architecture.

Below is our LeNet architechture with more structural details. You may check this figure below carefully before you start your LeNet implementation.

<img src="img/lenet-vert.png" width="200">

From the structual perspective, note that the height and width of the output of each layer throughout the convolutional block are reduced (compared with the previous layer):

The first convolutional layer uses 2 pixels of padding to compensate for the reduction in height and width that would otherwise result from using a $5 \times 5$ kernel.

The second convolutional layer forgoes padding, and thus the height and width are both reduced by 4 pixels. 

As we go up the stack of layers, the number of channels increases layer-over-layer from 1 in the input to 6 after the first convolutional layer and 16 after the second convolutional layer.

Meanwhile, each pooling layer halves the height and width.

Finally, each fully-connected layer reduces dimensionality, emitting an output whose dimension matches the total number of classes in our classification problem.

Now, let's build LeNet and train this CNN using PyTorch.

<font color='blue'>
    
### We provide you this file so that you can test and debug your task 3 implementations in your .py file by yourself, you don't need to modify any code in this jupyter notebook, but you are welcome to change implementations in this notebook if you want to play with this example model training process.
    
### You don't need to submit this notebook.
    
</font>

In [13]:
# use automatic reloading for your code from task3-template.py
# remember to rename task3-template.py before you submit it.
%reload_ext autoreload
%autoreload 2
%matplotlib notebook

In [14]:
import torch
from torch import nn

from task3_template import *

### 1. Load your LeNet-5 model

In [15]:
net = get_lenet()

In [16]:
X = torch.randn(size=(1, 1, 28, 28), dtype=torch.float32)
print("The output data shapes of all layers in your lenet are:")
for layer in net:
    X = layer(X)
    print(layer.__class__.__name__,'output shape: \t',X.shape)

The output data shapes of all layers in your lenet are:
Reshape output shape: 	 torch.Size([1, 1, 28, 28])
Conv2d output shape: 	 torch.Size([1, 6, 28, 28])
Sigmoid output shape: 	 torch.Size([1, 6, 28, 28])
AvgPool2d output shape: 	 torch.Size([1, 6, 14, 14])
Conv2d output shape: 	 torch.Size([1, 16, 10, 10])
Sigmoid output shape: 	 torch.Size([1, 16, 10, 10])
AvgPool2d output shape: 	 torch.Size([1, 16, 5, 5])
Flatten output shape: 	 torch.Size([1, 400])
Linear output shape: 	 torch.Size([1, 120])
Sigmoid output shape: 	 torch.Size([1, 120])
Linear output shape: 	 torch.Size([1, 84])
Sigmoid output shape: 	 torch.Size([1, 84])
Linear output shape: 	 torch.Size([1, 10])


In [17]:
print("The output data shapes of all layers in correct lenet are:")
print(get_correct_lenet_shape_str())

The output data shapes of all layers in correct lenet are:
Reshape output shape: 	 torch.Size([1, 1, 28, 28])
Conv2d output shape: 	 torch.Size([1, 6, 28, 28])
Sigmoid output shape: 	 torch.Size([1, 6, 28, 28])
AvgPool2d output shape: 	 torch.Size([1, 6, 14, 14])
Conv2d output shape: 	 torch.Size([1, 16, 10, 10])
Sigmoid output shape: 	 torch.Size([1, 16, 10, 10])
AvgPool2d output shape: 	 torch.Size([1, 16, 5, 5])
Flatten output shape: 	 torch.Size([1, 400])
Linear output shape: 	 torch.Size([1, 120])
Sigmoid output shape: 	 torch.Size([1, 120])
Linear output shape: 	 torch.Size([1, 84])
Sigmoid output shape: 	 torch.Size([1, 84])
Linear output shape: 	 torch.Size([1, 10])


#### In the two cells above, all the layer names, and the output data shapes of all the layers should be the same. 
#### (Please do not modify the layer names)

###  2. Train your LeNet model

In [18]:
batch_size = 256
train_iter, test_iter = load_data_fashion_mnist(batch_size=batch_size)

Now let us train and evaluate the LeNet-5 model using your implementation.

We recommend using Deep Learning accleration hardware (like CUDA compatible GPU). However, if you are using Mac or other computer without CUDA GPU, it is okay. After all, you only need to run LeNet training for 10 epochs.

If you are running without Deep Learning accleration hardware, we recommend that you print out test batch accuracy for every batch in every epoch in the training process, which might make things easier for debugging.

In [19]:
lr, num_epochs = 0.9, 10
# We recommend not to start training before you have got the correct LeNet-5 model
train_cnn(net, train_iter, test_iter, num_epochs, lr)

epoch 0




Test set loss:1.0593  Accuracy: 5962/10000 (59.62%)

epoch 1




Test set loss:0.8144  Accuracy: 6973/10000 (69.73%)

epoch 2




Test set loss:0.7269  Accuracy: 7282/10000 (72.82%)

epoch 3




Test set loss:0.6573  Accuracy: 7483/10000 (74.83%)

epoch 4




Test set loss:0.6285  Accuracy: 7571/10000 (75.71%)

epoch 5




Test set loss:0.5949  Accuracy: 7673/10000 (76.73%)

epoch 6




Test set loss:0.5598  Accuracy: 7858/10000 (78.58%)

epoch 7




Test set loss:0.5303  Accuracy: 8009/10000 (80.09%)

epoch 8




Test set loss:0.5162  Accuracy: 8020/10000 (80.20%)

epoch 9




Test set loss:0.5146  Accuracy: 8106/10000 (81.06%)




In [20]:
test_dataset_accuracy = evaluate_accuracy_try_gpu(net, test_iter)
print("After model training, your LeNet-5 achieves test dataset accuracy:",test_dataset_accuracy)

After model training, your LeNet-5 achieves test dataset accuracy: 0.8106


#### After 10 epochs of training with learning rate 0.9~1.1, batch size 256, a test dataset accuracy(on all 10000 test samples) >0.7000 is usually considered correct and will give you a full score for train_cnn function.

Congratulations on finishing task 3!